File: solutions/ruby/simple-cipher/1/simple_cipher.rb
class Cipher
  attr_reader :key

  def initialize(key='aaaaaaaaaa')
    raise ArgumentError.new() if key =~ /[^a-z]/ || key.empty?
    @key = key
  end

  def shift(char, key, &block)
    char_ord = char.ord - 'a'.ord
    key_ord = key.ord - 'a'.ord
    new_ord = block.call(char_ord, key_ord)
    (new_ord % 26 + 'a'.ord).chr
  end

  def encode(plaintext)
    cipher(plaintext, &:+)
  end

  def decode(plaintext)
    cipher(plaintext, &:-)
  end

  private
  def cipher(plaintext, &block)
    plaintext.chars.map.with_index{|c, index|
      key = @key[index % @key.length]
      shift(c, key, &block)
    }.join
  end
end

File: solutions/python/simple-cipher/1/simple_cipher.py
from string import ascii_lowercase
from typing import Literal
import random
from operator import add, sub


class Cipher:
    key = None

    def __init__(self, key=None):
        self.key = key or self.generate_random_key()

    @staticmethod
    def generate_random_key() -> str:
        return ''.join(random.choices(ascii_lowercase, k=100))

    @staticmethod
    def shift_char(char: str, key: str, mode: Literal['left', 'right']) -> str:
        operator = add if mode == 'right' else sub
        return chr((operator((ord(char) - 97), (ord(key) - 97)) % 26) + 97)

    @staticmethod
    def shift_string(text: str, key: str, mode: Literal['left', 'right']) -> str:
        retval = ''
        for index, char in enumerate(text):
            retval += Cipher.shift_char(char,
                                        key[index % len(key)], mode)
        return retval

    def encode(self, text: str) -> str:
        return self.shift_string(text, self.key, 'right')

    def decode(self, text):
        return self.shift_string(text, self.key, 'left')
